impl_core/src/main/java/io/opencensus/implcore/stats/RecordUtils.java

```diff
@@ -62,16 +61,12 @@
   @javax.annotation.Nullable @VisibleForTesting static final TagValue UNKNOWN_TAG_VALUE = null;
 
   static Map<TagKey, TagValue> getTagMap(TagContext ctx) {
-    if (ctx instanceof TagMapImpl) {
-      return ((TagMapImpl) ctx).getTags();
-    } else {
-      Map<TagKey, TagValue> tags = Maps.newHashMap();
-      for (Iterator<Tag> i = InternalUtils.getTags(ctx); i.hasNext(); ) {
-        Tag tag = i.next();
-        tags.put(tag.getKey(), tag.getValue());
-      }
-      return tags;
+    Map<TagKey, TagValue> tags = Maps.newHashMap();
```

Comment:
You force a new map creation every time this is sub-optimal because the case when the TagContext is not our implementation is very rare and probably we should move this conversion TagContext to TagMapImpl sooner and make our implementation work with TagMapImpl everywhere.

Then you use everywhere the Map<TagKey, ValueAndMetadata> to avoid to always create a new map.

So something like this:
```java
static TagMapImpl ToTagMapImpl(TagContext ctx) {
  if (!(ctx instanceof TagMapImpl)) {
    ...
    return  new TagMapImpl(...)
  }
  return (TagMapImpl) ctx
}

// In apis that we accept a TagContext like record we call this as a first thing
void record(TagContext ctx, MeasureMap) {
  TagMapImpl tagMap = ToTagMapImpl(ctx)
  // from here we only use TagMapImpl in our implementation.
}
```

Maybe this is an something we should do in a separate PR.

Reply:
Agreed non-native implementation should be rare. I reverted and modified this method to return `Map<TagKey, TagValueWithMetadata>` instead and avoid re-creating a map for native implementation. Though `TagMetadata` is not used in stats impl.